packages/flutter/lib/src/widgets/radio_base.dart

```diff
@@ -0,0 +1,207 @@
+  ///
+  /// This radio button is considered selected if its [value] matches the
+  /// [groupValue].
+  /// {@endtemplate}
+  final T? groupValue;
+
+  /// {@template flutter.widget.RadioBase.onChanged}
+  /// Called when the user selects this radio button.
+  ///
+  /// The radio button passes [value] as a parameter to this callback. The radio
+  /// button does not actually change state until the parent widget rebuilds the
+  /// radio button with the new [groupValue].
+  ///
+  /// If null, the radio button will be displayed as disabled.
+  ///
+  /// The provided callback will not be invoked if this radio button is already
+  /// selected.
+  ///
+  /// The callback provided to [onChanged] should update the state of the parent
+  /// [StatefulWidget] using the [State.setState] method, so that the parent
+  /// gets rebuilt.
+  /// {@endtemplate}
+  final ValueChanged<T?>? onChanged;
+
+  /// {@template flutter.widget.RadioBase.mouseCursor}
+  /// The cursor for a mouse pointer when it enters or is hovering over the
+  /// widget.
+  ///
+  /// If [mouseCursor] is a [WidgetStateMouseCursor],
+  /// [WidgetStateProperty.resolve] is used for the following [WidgetState]s:
+  ///
+  ///  * [WidgetState.selected].
+  ///  * [WidgetState.hovered].
+  ///  * [WidgetState.focused].
+  ///  * [WidgetState.disabled].
+  /// {@endtemplate}
+  ///
+  /// If null, then the value of [RadioThemeData.mouseCursor] is used.
+  /// If that is also null, then [WidgetStateMouseCursor.clickable] is used.
+  final WidgetStateProperty<MouseCursor> mouseCursor;
+
+  /// {@template flutter.widget.RadioBase.toggleable}
+  /// Set to true if this radio button is allowed to be returned to an
+  /// indeterminate state by selecting it again when selected.
+  ///
+  /// To indicate returning to an indeterminate state, [onChanged] will be
+  /// called with null.
+  ///
+  /// If true, [onChanged] can be called with [value] when selected while
+  /// [groupValue] != [value], or with null when selected again while
+  /// [groupValue] == [value].
+  ///
+  /// If false, [onChanged] will be called with [value] when it is selected
+  /// while [groupValue] != [value], and only by selecting another radio button
+  /// in the group (i.e. changing the value of [groupValue]) can this radio
+  /// button be unselected.
+  ///
+  /// The default is false.
+  /// {@endtemplate}
+  final bool toggleable;
+
+  /// The size of canvas for painter returned from [painterGetter] to paint.
+  final Size size;
+
+  /// {@macro flutter.widgets.Focus.focusNode}
+  final FocusNode? focusNode;
+
+  /// {@macro flutter.widgets.Focus.autofocus}
+  final bool autofocus;
+
+  /// The getter for the painter that used for painting the radio button.
+  ///
+  /// This getter is used when the build method is called. One is expected
+  /// to use the input `state` to update the painter before returning.
+  final RadioPainterGetter painterGetter;
```

Comment:
This API is this way because the UI requires the widget states in ToggeableStateMixin, such as when the state is hovered or selected.. etc.

An alternative is to make this a widget builder, but that will require changing the buildToggeable API.

Reply:
What do you think of the name `RadioPainterBuilder` and `painterBuilder` instead?

Reply:
I felt that builder implies return new object every time. The problem with this api is that CustomPainter is usually a changenotifier, so it has lifecycle. returning a new painter every time is wasteful and can easily forget to dispose.

Getter can be a permanent object that is why i go with this name that hopefully can convey more information.

though I think both are bad in this case.

An other alternative is to make Radio and CupertinoRadio extends RadioBase and RadioBaseState, that way RadioBaseState can leave an abstract painter getter, but I usually prefer composition vs inheritance for widgets

Reply:
> An other alternative is to make Radio and CupertinoRadio extends RadioBase

Being able to type check for a RadioBase could be useful for a lot of use cases for developers, especially package authors. I had assumed that would be needed for making the radio group widget work for both Material and Cupertino radios, but it looks like that might not be the case.

Reply:
right now both Radio and CupertinoRadio build a RadioBase with different config, and RadioGroup interacts with RadioBase.

I am going with the composition approach vs inheritance